src/GitHub.Exports/Extensions/VSExtensions.cs

```diff
@@ -33,20 +38,42 @@ public static object TryGetService(this IServiceProvider serviceProvider, Type t
 
         public static T GetService<T>(this IServiceProvider serviceProvider)
         {
-            return (T)serviceProvider.GetService(typeof(T));
+            if (cachedUIProvider != null && typeof(T) == typeof(IUIProvider))
+                return (T)cachedUIProvider;
+
+            return (T)GetServiceAndCache(serviceProvider, typeof(T), ref cachedUIProvider);
         }
 
         public static T GetExportedValue<T>(this IServiceProvider serviceProvider)
         {
+            if (cachedUIProvider != null && typeof(T) == typeof(IUIProvider))
+                return (T)cachedUIProvider;
+
             var ui = serviceProvider as IUIProvider;
             return ui != null
                 ? ui.GetService<T>()
-                : VisualStudio.Services.ComponentModel.DefaultExportProvider.GetExportedValue<T>();
+                : GetExportedValueAndCache<T, IUIProvider>(ref cachedUIProvider);
         }
 
         public static ITeamExplorerSection GetSection(this IServiceProvider serviceProvider, Guid section)
         {
             return serviceProvider?.GetService<ITeamExplorerPage>()?.GetSection(section);
         }
+
+        static object GetServiceAndCache<CacheType>(IServiceProvider provider, Type type, ref CacheType cache)
+        {
+            var ret = provider.GetService(type);
+            if (type == typeof(CacheType))
+                cache = (CacheType)ret;
+            return ret;
+        }
+
+        static T GetExportedValueAndCache<T, CacheType>(ref CacheType cache)
+        {
+            var ret = VisualStudio.Services.ComponentModel.DefaultExportProvider.GetExportedValue<T>();
+            if (typeof(T) == typeof(CacheType))
+                cache = (CacheType)(object)ret;
```

Comment:
The type system doesn't know that it's the same type at compile time inside this condition :wink: